Web: theme selector

stsrki opened the pull request:
I was a little bored, so I created a dark theme selector for our website. 

Currently, all it does is use a default dark theme that comes from Bootstrap 5 by default. I have implemented a simple ThemeService that just raises a `changed` event when the theme is selected, and then in various places, I listen for the event. Then, just apply light or dark colors where needed.

David-Moreira commented:
> I was a little bored, so I created a dark theme selector for our website.

Little bored? There's plenty of tickets to go around!! (j/k j/k :D)

stsrki commented:
> > I was a little bored, so I created a dark theme selector for our website.
> 
> Little bored? There's plenty of tickets to go around!! (j/k j/k :D)

I'm leaving those as your last dance 😜

David-Moreira commented on Documentation/Blazorise.Docs/App.razor:
```diff
@@ -117,3 +118,20 @@
+
+    protected override void OnInitialized()
+    {
+        ThemeService.ThemeChanged += OnThemeChanged;
```
Not sure whether dispose matters or not here?

David-Moreira commented on Documentation/Blazorise.Docs/Components/Commercial/PlanSwitch.razor.cs:
```diff
@@ -0,0 +1,43 @@
+
+public partial class PlanSwitch : IDisposable
+{
+    private Background GetSelectedBackground( string value ) => SelectedValue == value ? Background.Primary : Background.Default;
```
I don't understand these getters, if the **SelectedValue** is already equal to the value that's in the argument it chooses Background.Primary, otherwise Default?

stsrki replied:
It just applies the primary color to the selected button. That's it.

David-Moreira replied:
I still don't understand this at all personally... But as long as it's working as expected, I'm ok with it, then.

David-Moreira commented on Documentation/Blazorise.Docs/Components/Commercial/PlanSwitch.razor.cs:
```diff
@@ -0,0 +1,43 @@
+{
+    private Background GetSelectedBackground( string value ) => SelectedValue == value ? Background.Primary : Background.Default;
+
+    private TextColor GetTextSelectedColor( string value ) => SelectedValue == value
```
Same thing, don't understand why SelectedValue == value turns into TextColor.White?

stsrki replied:
White looks better on the primary purple color. That's why it is white in selected mode.

David-Moreira commented on Documentation/Blazorise.Docs/Components/Footer.razor:
```diff
@@ -103,4 +104,26 @@
+        ThemeService.ThemeChanged -= OnThemeChanged;
+    }
+
+    async void OnThemeChanged( object sender, string theme )
```
If you do StateHasChanged in app.razor, is that not enough to update all the underlying components?

stsrki replied:
I have removed code from App.razor. It didn't work

David-Moreira commented on Documentation/Blazorise.Docs/Routes.razor.cs:
```diff
@@ -0,0 +1,56 @@
+        base.OnInitialized();
+    }
+
+    protected override async void OnAfterRender( bool firstRender )
```
Am I missing something? Isn't there a trully async version?

David-Moreira commented on Documentation/Blazorise.Docs/Components/Footer.razor:
```diff
@@ -103,4 +104,26 @@
+@code{
+    [Inject] private ThemeService ThemeService { get; set; }
+
+    private TextColor IconTextColor => ThemeService.CurrentTheme == "Dark" ? TextColor.Light : TextColor.Dark;
```
`ThemeService.IsDark` helper?

David-Moreira commented on Documentation/Blazorise.Docs/Routes.razor.cs:
```diff
@@ -0,0 +1,56 @@
+
+    async void OnThemeChanged( object sender, string theme )
+    {
+        if ( theme == "Dark" )
```
Check some centralized constant on `ThemeService`?

David-Moreira commented on Documentation/Blazorise.Docs/Services/ThemeService.cs:
```diff
@@ -0,0 +1,33 @@
+    public Background BarBackground = Background.Light;
+    public ThemeContrast BarThemeContrast = ThemeContrast.Light;
+
+    public bool IsDark => CurrentTheme == "Dark";
```
"Dark" as constant?

David-Moreira commented on Source/Blazorise/wwwroot/utilities.js:
```diff
@@ -24,6 +24,14 @@ export function toggleClass(element, classname) {
 }
 
+export function addAttribute(element, attribute, value) {
+    element.setAttribute(attribute, value);
```
You usually have the habit of null checking? Should null check?

tesar-tech commented on Documentation/Blazorise.Docs/Routes.razor.cs:
```diff
@@ -0,0 +1,56 @@
+            await JSUtilitiesModule.RemoveAttributeFromBody( "data-bs-theme" );
+        }
+
+        await LocalStorage.SetItemAsync( "theme", theme );
```
Using local storage with Blazor Server results in unintended behavior. 
Since the first "render" happens on the server, where local storage is not accessible, it will always serve the light (default) mode. Then, it checks for the "dark" mode and switches accordingly. This results in a quick "light theme blink" on refresh, which is exactly what's happening. The "quickness" of the blink depends on internet speed. You can throttle in devtools for a more noticeable effect. 

To my knowledge, the only solution is to store the "dark" mode value inside a cookie, which will be sent together with the first request. This ensures the first render happens in the correct dark/light theme.

stsrki replied:
That's the limitation at this time. We could use cookies but the GDPR consent complicates it. SO I would just leave it as it is and go with the flicker.

tesar-tech commented on Documentation/Blazorise.Docs/Components/DocsAttributes.razor.cs:
```diff
@@ -1,16 +1,30 @@
 
+    protected override void OnInitialized()
+    {
+        ThemeService.ThemeChanged += OnThemeChanged;
```
This is here because we want to make sure the page re-render with the correct theme (`@ThemeContrast`)? 
I think it's enough if the re-render happens in some parent component, resulting in re-render also in here, thus the whole `ThemeService.ThemeChanged`, `StateHasChanged` code has no use in here (and in other child components). It just adds another re-render.

stsrki replied:
It doesn't work all the times if I leave it at parent level. That's why I have the event handler at child components.

tesar-tech replied:
I see.  In that case I would suggest to investigate the reason why it doesn't work, because it doesn't appear as standard behavior and may backfire badly in the future... 
Also doing the event subscription and un-subscription in every component where you want to use a `ThemeVariable` looks quite painful to me...

My guess is that something is wrong with the non-async version of the ThemeChanged (` async void OnThemeChanged`).. Maybe making it to return Task<string> would solve the issue (really, just a guess).

tesar-tech commented on Documentation/Blazorise.Docs/Routes.razor.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            var theme = await LocalStorage.GetItemAsync<string>( "theme" );
+
+            if ( ThemeService.CurrentTheme != theme )
```
What about putting the check (`( ThemeService.CurrentTheme != theme )`) inside the `SetTheme`?
This will also "protect" from the needless re-renders when user with light theme clicks again the light theme..

tesar-tech commented on Documentation/Blazorise.Docs/wwwroot/blazorise.docs.min.css:
my favorite topic - generated file should not be in vcs.

stsrki replied:
As mentioned before. CSS files must be included.

tesar-tech reviewed (commented):
For the dark/light theme switcher, we should also implement a third value, "use system default," which should be the default option.
 See [this](https://jonhilton.net/blazor-tailwind-dark-mode-local-storage/) for implementation (scroll down to `Respect their OS preferences`).

stsrki commented on Documentation/Blazorise.Docs/Styles/_base.scss:
```diff
@@ -9,8 +12,28 @@
     -moz-osx-font-smoothing: grayscale;
 }
 
+body {
```
This file must be present at all times. SCSS is a compiler with the help of VS extension during the development stage. So we need to keep the final CSS files.

stsrki commented:
I have fixed a bunch of code according to your comments. And introduced the new System option to detect dark mode.

It's not the most beautiful code but it works. Please retest.

David-Moreira commented on Documentation/Blazorise.Docs/App.razor:
```diff
@@ -117,3 +117,6 @@
     </script>
 </body>
 </html>
+@code {
+    RenderFragment customIcon =@<Image Source="/img/logos/blazorise-small.svg" Text="Small Blazorise logo" Style="width:32px; height: 32px" />;
+}
```
Seems like it's added and not used can you check?

David-Moreira commented on Documentation/Blazorise.Docs/Services/ThemeService.cs:
```diff
@@ -0,0 +1,54 @@
+
+namespace Blazorise.Docs.Services;
+
+public class ThemeService
```
Make sure that the accessors properly encapsulate logic, just for good practices sake, if you care about it, otherwise it's fine.
i.e **SystemIsDarkMode** can be publicly set directly.

David-Moreira commented on Source/Blazorise/Interfaces/Modules/IJSUtilitiesModule.cs:
```diff
@@ -158,4 +173,10 @@ public interface IJSUtilitiesModule : IBaseJSModule
+    /// Checks if the current theme is in dark mode.
+    /// </summary>
+    /// <returns>A task that represents the asynchronous operation. The task result contains true if the theme is in dark mode, otherwise false.</returns>
+    ValueTask<bool> IsDarkMode();
```
Since this is an utilities module, we could potentially make it clear that It's related to the Theme, i.e **IsThemeDarkMode**

David-Moreira replied:
Actually, maybe even `IsSystemDarkMode` just to be clear, that this is never user set. but whatever it is figured out from the browser/system.
